### PR TITLE
(fixes #13) Support SQL related databases and the relevant data nodes

### LINKED ISSUE #13: Support SQL related databases and the relevant data nodes
- http://docs.aws.amazon.com/datapipeline/latest/DeveloperGuide/dp-object-rdsdatabase.html
- http://docs.aws.amazon.com/datapipeline/latest/DeveloperGuide/dp-object-sqldatanode.html

### PR DESCRIPTION
**Pull request by realstraw**

### CODE REVIEW

File: src/main/scala/com/krux/hyperion/objects/CopyActivity.scala [hunk 1]
```diff
@@ -1,17 +1,33 @@
...
+ */
+case class CopyActivity private (
   id: String,
-  input: DataNode,
-  output: DataNode,
+  input: Copyable,
+  output: Copyable,
   runsOn: Ec2Resource,
-  dependsOn: Seq[PipelineActivity] = Seq()
+  dependsOn: Seq[PipelineActivity]
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/CopyActivity.scala]: Why does this have to have a `dependsOn`?

Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/CopyActivity.scala]: Ah, nm.  I see the companion object now.

File: src/main/scala/com/krux/hyperion/objects/Copyable.scala [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package com.krux.hyperion.objects
+
+trait Copyable extends DataNode
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/Copyable.scala]: Why the additional trait?

File: src/main/scala/com/krux/hyperion/objects/CsvDataFormat.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * CSV data format
+ */
+case class CsvDataFormat private (
+    id: String,
+    column: Option[Seq[String]] = None
+  ) extends DataFormat {
+
+  def withColumns(cols: Seq[String]) = this.copy(column = Some(cols))
+  def serialize = AdpCsvDataFormat(id, Some(id), column, None)
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/CsvDataFormat.scala]: Put the parameters on another line and prefix with the parameter name for consistency.

File: src/main/scala/com/krux/hyperion/objects/Ec2Resource.scala [hunk 4]
```diff
@@ -66,7 +75,7 @@ case class Ec2Resource(
 
 object Ec2Resource {
   def apply()(implicit hc: HyperionContext) = new Ec2Resource(
-    id = "Ec2Resource",
+    id = PipelineId.generateNewId("Ec2Resource"),
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/Ec2Resource.scala]: This seems unrelated to this change.  Why?

File: src/main/scala/com/krux/hyperion/objects/MySqlDatabase.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  def connectionString: String
+
+  def jdbcDriverClass: String
+
+  def serialize = AdpJdbcDatabase(
+      id = id,
+      name = Some(id),
+      connectionString = connectionString,
+      jdbcDriverClass = jdbcDriverClass,
+      databaseName = None,
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/MySqlDatabase.scala]: Why hardcode these as None?

File: src/main/scala/com/krux/hyperion/objects/SparkActivity.scala [hunk 6]
```diff
@@ -19,7 +19,7 @@ case class SparkActivity (
   def dependsOn(activities: PipelineActivity*) = this.copy(dependsOn = activities)
   def forClient(client: String) = this.copy(id = s"${id}_${client}")
 
-  override def objects: Iterable[PipelineObject] = Seq(runsOn) ++ dependsOn
+  override def objects: Iterable[PipelineObject] = runsOn +: dependsOn
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/SparkActivity.scala]: Nice!

File: src/main/scala/com/krux/hyperion/objects/SqlActivity.scala [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  def serialize = AdpSqlActivity(
+      id = id,
+      name = Some(id),
+      database = AdpRef[AdpDatabase](database.id),
+      script = script,
+      scriptArgument = scriptArgument match {
+        case Seq() => None
+        case other => Some(other)
+      },
+      queue = None,
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/SqlActivity.scala]: Why hardcode `None`?

File: src/main/scala/com/krux/hyperion/objects/aws/AdpDataFormats.scala [hunk 8]
```diff
@@ -32,3 +32,13 @@ case class AdpTsvDataFormat(
...
 }
+
+
+case class AdpCsvDataFormat(
+    id: String,
+    name: Option[String],
+    column: Option[Seq[String]],
+    escapeChar: Option[String]
+  ) extends AdpDataFormat {
+  val `type` = "CSV"
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/aws/AdpDataFormats.scala]: Newline before/after for consistency.

File: src/main/scala/com/krux/hyperion/objects/aws/AdpDataNodes.scala [hunk 9]
```diff
@@ -108,3 +108,31 @@ case class AdpRedshiftDataNode(
...
+    id: String,
+    name: Option[String],
+    table: String,
+    username: String,
+    `*password`: String,
+    connectionString: String,
+    selectQuery: Option[String],
+    insertQuery: Option[String]
+  ) extends AdpDataNode {
+  val `type` = "SqlDataNode"
```
Comment (sethyates) [inline src/main/scala/com/krux/hyperion/objects/aws/AdpDataNodes.scala]: Newline before/after for consistency.

**Comment by sethyates** [comment]
Mostly nits.  I'm :+1: on the Database changes (but I'd like to see all options passed up instead of `None` hardcoded); however, I'm :-1: on mixing in a (partial?) solution for #17 at the same time.

File: src/main/scala/com/krux/hyperion/objects/Ec2Resource.scala (hunk 4 above)
Comment (realstraw) [inline src/main/scala/com/krux/hyperion/objects/Ec2Resource.scala]: I'll revert this back and put it in another PR

File: src/main/scala/com/krux/hyperion/objects/MySqlDatabase.scala (hunk 5 above)
Comment (realstraw) [inline src/main/scala/com/krux/hyperion/objects/MySqlDatabase.scala]: Was being lazy, I'll add it.

File: src/main/scala/com/krux/hyperion/objects/Copyable.scala (hunk 2 above)
Comment (realstraw) [inline src/main/scala/com/krux/hyperion/objects/Copyable.scala]: Because the CopyActivity supports only S3DataNode and SqlDataNode, not other DataNodes

**Comment by realstraw** [comment]
@sethyates done

**Comment by sethyates** [comment]
LGTM :sparkles: :sparkles: :sparkles: (make sure you squash on merge)

**Comment by realstraw** [comment]
merged by c5a7a24b1bac193a94e90b4eae0d075edf670520
